Fix ReconnectingCli deadlock on destructor

marosmars opened the pull request:
For some reason, shared_from_this in thenError callback prevented
the destructor to release the executor...

Using raw pointer instead, with params as protection.

+ Use proper ex type when asserting exception in tests

Signed-off-by: Maros Marsalek <mmarsalek@frinx.io>

tomasol commented:
Thanks for the fix, however when I run ReconnectingCliTest, I still see hang. I removed shared_from_this completely as it was my mistake not to do so previously. Please check my commit here:  https://github.com/tomasol/magma/pull/new/reconnect-deadlock-fix
This also fixed the deadlock in test for me.

hotlib commented on devmand/gateway/src/devmand/test/cli/PromptAwareCliTest.cpp:
```diff
@@ -50,16 +50,19 @@ using namespace devmand::test::utils::ssh;
  protected:
   shared_ptr<CPUThreadPoolExecutor> testExec;
+  shared_ptr<server> ssh;
 
```
is ssh needed without the promptawarecli test?

hotlib reviewed (approved):
ok

marosmars committed:
Add missing semi() + fix tests

+ Fix keepalives being filtered by cache
+ Remove unnecessary tests
+ Fix PromptAwareCli race condition between destructor and callback
+ disable DEBUG logs for QueuedCli test

Signed-off-by: Maros Marsalek <mmarsalek@frinx.io>

marosmars commented:
> Thanks for the fix, however when I run ReconnectingCliTest, I still see hang. I removed shared_from_this completely as it was my mistake not to do so previously. Please check my commit here: https://github.com/tomasol/magma/pull/new/reconnect-deadlock-fix
> This also fixed the deadlock in test for me.

OK. I changed this PR so that it contains:

- explicit call of semi() where missing ... we relied on cast from Future to SemiFuture (whatever that does)
- Finish keepalive test in ReconnectingSshTest by verifying what was received on server
- Fix QueuedCli test to not fail
- Remove redundant tests

tomasol commented on devmand/gateway/src/devmand/channels/cli/TimeoutTrackingCli.cpp:
```diff
@@ -71,9 +71,13 @@ folly::SemiFuture<std::string> TimeoutTrackingCli::executeRead(
+  return executeSomething(
+             cmd,
+             "TTCli.executeWrite",
+             [this, cmd]() {
```
Capturing `this` in lambdas is problematic (in general). Can we capture timeoutTrackingParameters instead?

marosmars replied:
Yes will do... 

This is just a reformat of existing code + I think the lambda is called from a protected context inside executeSomething

tomasol replied:
ok agree.
